File: erettsegi/emelt/2010_may/python/Helyjegy.py
from itertools import chain

class Utas:

    def __init__(self, line: str, sorszam: int):
        split = line.split(' ')

        self.sorszam = sorszam
        self.ules = int(split[0])
        self.felszallas_km = int(split[1])
        self.leszallas_km = int(split[2])

def get_ar(utas: Utas, ar_per_10km: int):
    tav = utas.leszallas_km - utas.felszallas_km
    utolsoSzamjegy = tav % 10
    tizesek = tav // 10 + (1 if(utolsoSzamjegy == 3 or utolsoSzamjegy == 4 or utolsoSzamjegy == 8 or utolsoSzamjegy == 9) else 0)

    return ar_per_10km * tizesek

def get_ules_status(ules: int, bekert_km: int, utasok: list[Utas]):
    ules_stat = 'üres'

    for utas in utasok:
        if utas.ules == ules and (utas.felszallas_km == bekert_km or utas.leszallas_km == bekert_km):
            ules_stat = f'{utas.sorszam}. utas'
            break

    return str(ules) + '. ülés: ' + ules_stat + '\n'

with open('eladott.txt') as input_file:
    lines = input_file.readlines()
    first_line_split = lines[0].split(' ')
    vonal_hossz = int(first_line_split[1])
    ar_per_10km = int(first_line_split[2])
    utasok = [ Utas(lines[i], i) for i in range(1, len(lines)) ]

utolso = utasok[-1]

print(f'2.Feladat: Utolsó utas ülése: {utolso.ules} utazott távolság: {utolso.leszallas_km - utolso.felszallas_km}')
print('3.Feladat:')

for utas in utasok:
    if (utas.leszallas_km - utas.felszallas_km) == vonal_hossz:
        print(utas.sorszam, end = ' ')

print(f'\n4. Feladat: Összes bevétel: {sum(get_ar(k, ar_per_10km) for k in utasok)}')

utolso_elotti_megallo_km = max(k.leszallas_km for k in utasok if k.leszallas_km != vonal_hossz)
utoljara_felszallok = sum(1 for k in utasok if k.felszallas_km == utolso_elotti_megallo_km)
utoljara_leszallok  = sum(1 for k in utasok if k.leszallas_km == utolso_elotti_megallo_km)

print(f'5.Feladat: Utolsó megállónál felszállók: {utoljara_felszallok}, leszállók: {utoljara_leszallok}')

osszes_megallo = chain((k.leszallas_km for k in utasok), (k.felszallas_km for k in utasok))
megallok_szama = len(set(osszes_megallo)) - 2

print(f'6.Feladat: Megállók száma: {megallok_szama}')

bekert_km = int(input('Írj be 1 km számot!'))
fileba = (get_ules_status(i, bekert_km, utasok) for i in range(1, 49))

with open('kihol.txt', 'w') as output:
    output.writelines(fileba)

File: erettsegi/emelt/2010_may/rust/Helyjegy.rs
use std::fs::File;
use std::io::{ BufReader, BufRead, stdin, Write };

use itertools::Itertools;

struct Utas {
    sorszam: usize,
    ules: i32,
    felszallas_km: i32,
    leszallas_km: i32
}

fn create_utas(line: &str, sorszam: usize) -> Utas {
    let split = line.split(' ').collect::<Vec<_>>();

    Utas {
        sorszam: sorszam,
        ules: split[0].parse::<i32>().unwrap(),
        felszallas_km: split[1].parse::<i32>().unwrap(),
        leszallas_km: split[2].parse::<i32>().unwrap()
    }
}

fn get_ar(utas: &Utas, ar_per_10km: i32) -> i32 {
    let tav = utas.leszallas_km - utas.felszallas_km;
    let utolso_szamjegy = tav % 10;
    let tizesek = tav / 10 + (if utolso_szamjegy == 3 || utolso_szamjegy == 4 || utolso_szamjegy == 8 || utolso_szamjegy == 9 { 1 } else { 0 });

    ar_per_10km * tizesek
}

fn get_ules_status(ules: i32, bekert_km: i32, utasok: &Vec<Utas>) -> String {
    let ules_status = utasok.iter()
                            .filter(|k| k.ules == ules)
                            .find(|k| k.felszallas_km == bekert_km || k.leszallas_km == bekert_km)
                            .map(|k| format!("{}. utas", k.sorszam))
                            .unwrap_or("üres".to_string());

    return format!("{}. ülés: {}", ules, ules_status);
}

fn main() {
    let lines = BufReader::new(File::open("eladott.txt").unwrap())
                         .lines()
                         .map(|k| k.unwrap())
                         .collect::<Vec<_>>();

    let first_line_split = lines[0].split(' ').collect::<Vec<_>>();
    let vonal_hossz = first_line_split[1].parse::<i32>().unwrap();
    let ar_per_10km = first_line_split[2].parse::<i32>().unwrap();
    let utasok = (1 .. lines.len()).map(|i| create_utas(&lines[i], i))
                                   .collect::<Vec<_>>();

    let utolso = &utasok[utasok.len() - 1];

    println!("2.Feladat: Utolsó utas ülése: {} utazott távolság: {}", utolso.ules, utolso.leszallas_km - utolso.felszallas_km);
    println!("3.Feladat:");

    utasok.iter()
          .filter(|k| (k.leszallas_km - k.felszallas_km) == vonal_hossz)
          .for_each(|k| print!("{} ", k.sorszam));

    println!("\n4.Feladat: Összes bevétel: {}", utasok.iter().map(|k| get_ar(k, ar_per_10km)).sum::<i32>());

    let utolso_elotti_megallo_km = utasok.iter()
                                         .map(|k| k.leszallas_km)
                                         .filter(|k| *k != vonal_hossz)
                                         .max()
                                         .unwrap();

    let utoljara_felszallok = utasok.iter().filter(|k| k.felszallas_km == utolso_elotti_megallo_km).count();
    let utoljara_leszallok  = utasok.iter().filter(|k| k.leszallas_km == utolso_elotti_megallo_km).count();

    println!("5.Feladat: Utolsó megállónál felszállók: {}, leszállók: {}", utoljara_felszallok, utoljara_leszallok);

    let megallok_szama = utasok.iter().map(|k| k.leszallas_km)
                               .chain(utasok.iter().map(|k| k.felszallas_km))
                               .unique()
                               .count() - 2;

    println!("6.Feladat: Megállók száma: {}", megallok_szama);
    println!("Írj be 1 km számot!");

    let mut bekert_km_str = String::new();
    stdin().read_line(&mut bekert_km_str);

    let bekert_km = bekert_km_str.trim_end().parse::<i32>().unwrap();
    let fileba = (1 .. 49).map(|i| get_ules_status(i, bekert_km, &utasok)).join("\n");

    write!(File::create("kihol.txt").unwrap(), "{}", fileba);
}